### DISCUSSION TITLE
Assembler optimizations

### DISCUSSION THREAD

**Discussion by puzzledsab** [sabnzbd/sabnzbd#2398] [General]
I've started this discussion for more general talk about Python and file assembler optimizations. The 16 KB SSL discussion is getting a bit long and broad and it seems better to keep #2396 specifically for the technical issues of replacing MD5 with CRC32.

I have done some isolated testing of 750 KB blocks while calculating MD5. My conclusion is that it can be parallelized fairly well. Even with just one thread, the calculating and writing overlaps somewhat. It's probably because of the write cache. Doing MD5 in a separate thread can improve it to be more or less fully parallelized. In my test branch I've put it in a separate thread: https://github.com/sabnzbd/sabnzbd/pull/2391/files

We could even run multiple MD5 threads as long as most usenet posts consists of lots of RAR files by downloading blocks from multiple files simultaneously. If we have x threads with their own queue then we could attach threads with queues to each file (nzf) in a round robin fashion. It would make sure SAB could scale better as CPUs get more cores without increasing the frequency much. It wouldn't work on spinning disks but those probably won't be faster than single core MD5 calculations anyway.

There was some discussion about sparse files. I'm not sure if that's a good idea, for the same reason. I assume that if we write parts to random parts in a file then the write speed will go down on a spinning drive. I suspect that this may sometimes be the reason why some people get a higher speed with SAB than Nzbget.

@mnightingale: Regarding your last message, I think that in theory it shouldn't be necessary to release the GIL by using sleep. It happens anyway both when writing data and calculating MD5. Because of this I think it's already releasing the GIL most of the time it's running. If you want to try without the unnecessary loops and file openings you can apply this change: https://github.com/sabnzbd/sabnzbd/pull/2392/files

**Comment by animetosho** [comment]
If you want to stick with MD5, I'd recommend ignoring the file MD5 and try to compute block based MD5s instead.  Block MD5s won't be less reliable than a file MD5 (unlike CRC32).  
(I've argued that PAR2's file-based MD5 is rather useless and only limits performance; a PAR2 writer can't avoid computing it, due to the spec, but a PAR2 reader can easily pretend the file MD5 doesn't exist)

> We could even run multiple MD5 threads as long as most usenet posts consists of lots of RAR files by downloading blocks from multiple files simultaneously

The assumption obviously breaks if the upload isn't a bunch of parts. But as you point out, you generally want to write sequentially as much as possible, without jumping across files.  
I [implemented a strategy](https://github.com/Yutaka-Sawada/MultiPar/issues/21#issuecomment-971571710) for balancing MD5 parallelism and sequential I/O during PAR2 creation. If you *really* want to stick with file-based MD5, it might be worth considering - it may work better than round-robin scheduling for spinning disks (probably doesn't matter much for SSDs).

But even in the ideal scenario, it scales less than what a block-based MD5 approach can achieve, hence my recommendation to avoid relying on file-based MD5s if possible.

> There was some discussion about sparse files. I'm not sure if that's a good idea, for the same reason. I assume that if we write parts to random parts in a file then the write speed will go down on a spinning drive.

Sparse/pre-allocated files doesn't prevent you from writing sequentially though. They just provide the *option* to perform a random write if necessary (e.g. missing articles, article cache fills up etc). Other benefits are [listed here](https://github.com/sabnzbd/sabnzbd/issues/2169#issuecomment-1119111265).

**Reply by puzzledsab** [reply]
> If you want to stick with MD5, I'd recommend ignoring the file MD5 and try to compute block based MD5s instead. Block MD5s won't be less reliable than a file MD5 (unlike CRC32). (I've argued that PAR2's file-based MD5 is rather useless and only limits performance; a PAR2 writer can't avoid computing it, due to the spec, but a PAR2 reader can easily pretend the file MD5 doesn't exist)

Just to be clear, I'm very grateful for all your suggestions. There has been such a rush of big optimization ideas the last few days that it's a bit overwhelming. I will certainly look into this eventually if nobody else does it first.

Perhaps the main advantage of the file MD5 is that it's very simple. It can be used both to identify the file in the PAR2 set without knowing the block size and to verify the data. There have been a few cases where people put par2 files in rar files which can be used to deobfuscate the name after the download is complete. Those may not use the same part size. Of course, we could use the 16KB block for this.

> Sparse/pre-allocated files doesn't prevent you from writing sequentially though. They just provide the _option_ to perform a random write if necessary (e.g. missing articles, article cache fills up etc). Other benefits are [listed here](https://github.com/sabnzbd/sabnzbd/issues/2169#issuecomment-1119111265).

I can certainly see your point and in theory I don't think it should be all that hard to implement where it is supported. My main worry right now is how to determine when it can be used safely and when it would not work or degrade performance, for instance because it would allocate a huge file and fill it with zeros.

**Reply by animetosho** [reply]
Apologies if my comment came across as rather strong or inconsiderate - I mostly wanted to just start some discussion. I think your ideas + contribution has a lot of merit, particularly with its simplicity (so much so that you've already implemented it!).

I was focusing more on if *further* improvements are desired. If it's all too much, I completely understand and have no qualms about throwing it on the 'maybe think about it later' pile.

> There have been a few cases where people put par2 files in rar files which can be used to deobfuscate the name after the download is complete

(rant) Uploaders really do like doing some of the silliest things... I really wish they'd cut down on the shenanigans, though I sometimes wonder whether I'm actually the crazy one.  
PAR2 is designed to be resilient against corruption. That obviously falls apart if you do something as questionable as wrap it in a RAR that could fail to extract.

...anyway, if the PAR2 files are in a RAR:

* if it's in the same archive as the other files, then Quick Check just won't work, since everything needs to be extracted to check anything, and the PAR2 is much less useful if the files extracted without issue
* if it's in a separate archive, Sabnzbd would need to know which archive to extract to get the PAR2, so that it can check. In such a case, downloading and extracting the archive as the first step in the download process sounds feasible, though may add complexity
* if Sabnzbd can't tell which archive contains the PAR2, and just tries brute forcing all archives, it'll know if something is corrupt due to the extract failing. In such a case, Quick Check is much less useful (if at all).
* if the uploader supplies *multiple* PAR2 sets, with different block sizes, I suppose the only solution is to use the 16KB hash. With file-based MD5, you have the advantage of using that instead of the 16KB hash, so there's a benefit I guess, though it is more likely to break in the case of missing articles

Of course, the complexity with block-based MD5 is that the PAR2 needs to be downloaded first (or at least close enough) to obtain the relevant block size.

> My main worry right now is how to determine when it can be used safely and when it would not work or degrade performance, for instance because it would allocate a huge file and fill it with zeros.

It should be fine on the most widely used filesystems out there (may have to check with the FAT family). Not sure about network attached storage - may need checking on that.

Most applications I've seen just expose it as an option, with preallocation enabled by default. If you want to stay cautious, you could just leave it disabled at first.

**Comment by mnightingale** [comment]
Duplicating from the 16 KB discussion and a few additional things from some observations I found of the assembler behaviour:

- My PC seems to be able to MD5 at around about 1 GB/s this has the consequence of bottlenecking the maximum speed the assembler can write at
- From this I observed that if the MD5 calculation was no longer required (I hardcoded it and commented out the update/digest calls) or replaced with a quicker calculation (CRC) and the bottleneck removed the result would be that the article cache can be flushed to disk more often - this is likely also the path most users experience currently, if they are downloading at say 50 MB/s then the assembler could be ran thousands of times using more CPU then if it could batch them together better:
  - I found 10 GB downloading at ~2 GB/s called the assembler ~250 times
  - The logic is roughly:
    -  [IO] open nzf.filepath
    - for each article
    - continue if article has already been written
    - [IO] write article
    - [CPU] update MD5
    - [IO] close nzf.filepath
  -  The actual path could be [IO] open => [IO] write very few files => [IO] close
  - Or in my case above with a 10 GB download assembled 250 times only write 40MB per iteration on average
  - I think it's expensive to open => write a small amount of data => close
  - Larger sequential writes can be written faster on SSD and HDD

My PC and in-memory testing setup is quite extreme however, if optimisations can be found allowing faster speeds to be possible then it will also have the effect of reducing CPU usage for slower connections or better performance on slower devices.

If any system can download faster than it can MD5 or I suppose faster than it can write the article cache then the article cache can become full, at which point it is bypassed and writes straight to disk which the assembler than needs to read again.
I think this aspect probably warrants a separate discussion.

I think there are a couple of areas we should consider optimisations to:
- The assembler IO open/close, perhaps don't use "with" and only open if something needs writing
  - Maybe the file could just be kept open - throughout the assemble state or for a time limit of inactivity.
- Logic around determining if it's worth the assembler running
  - On the 16 KB discussion I mentioned testing a naïve approach of just sleeping after every assemble, I agree with @animetosho that wouldn't be a good design
  - As @animetosho proposed maybe query the article cache length - article cache size could be 0 so something to be aware of.
  - @animetosho mentions write and writev, Python does have os.writev (unix only), but I would expect the high level file handling in Python is already handling that for us
  - Looping through all the articles to find one that hasn't been written #2392 - I don't think the looping is very expensive, I see @puzzledsab has since revised it to only open the file if needed - instead I propose a design were `nzf.decodetable` is a queue and pops it if `article.on_disk`.
  - I think maybe a calculation related to article cache utilisation and/or how long an item has been cached or downloaded for, e.g. if the "next" article was downloaded >1 second ago then run the assembler consuming the cache and any articles that were downloaded in the last second.
  - May need to consider SSD vs HDD performance
 
Please share any thoughts on the above or your own ideas for the optimal "should the assembler run" logic.

**Reply by thezoggy** [reply]
wouldnt we benefit by moving our decoder/downloader to use generators so we could iterate and reduce memory?

**Reply by mnightingale** [reply]
I'm not sure how much of the above is relevant anymore but some of it will be and as I get more familiar with the codebase I think some areas would benefit from different datastructures, it uses lists a lot when sets or deques would be a better fit.

It all gets a bit complicated because some of the calls also have sideeffects but I do agree that generators or dequeues would be beneficial, a couple of areas I've noticed:
- The downloader maintains a list of the next 20 articles it's going to attempt so it involves the creation of lots of new lists, it eventually gets back to NzbFile.articles which stores all the pending articles which isn't an ideal use of lists - I think this is a good candidate for a generator and avoid storing the next 20 articles. 
- As mentioned above the NzbFile.decodetable also stores a list of a files articles for the assembler, but the assembler can end up looping through the same articles many times until it gets to one it hasn't written to disk yet, it's not significan't memory usage because it's just metadata about an aritcle (id, size, etc.) - I'm not sure of a better way to structure this currently but I also don't think assembly in order is always required anyway for example if writing to an SSD

They're all pretty small wins and probably insignicant compared to the time spend reading sockets or decoding but I do think there are improvements to be made.
I'd start with profiling the code in the downloader/assembler at 'normal' speeds i.e. limit to <100MB/s and see which functions/lines get called a lot.

**Reply by thezoggy** [reply]
isnt the point of the article cache is to take in the decoded stuff and write the file in order?

**Reply by animetosho** [reply]
Previously, MD5 was computed, and MD5 must be computed in order. Now that's no longer being performed, assembling in order isn't strictly required.

Sequential writing is generally preferred, even on SSDs, so having an article cache makes sense, in my opinion. However, the current strategy of spilling to disk, when cache space is depleted, is worse than just doing an in-place out-of-order write (i.e. skipping ahead and writing); the spill isn't going to be a sequential write anyway (since it's to a separate file), so either way, you're doing an out-of-order write, but spilling has the added cost of needing to eventually re-read and write the article again.

This would also allow the article cache to be more modestly sized (<100MB), and eliminates the notion of temporary article storage altogether.

**Reply by mnightingale** [reply]
> isnt the point of the article cache is to take in the decoded stuff and write the file in order?

Yes because sequential writes are faster requring fewer diskoperations and for HDDs it should result in less fragmented files.

After taking a look I don't think the assembler looping logic is a problem because it waits until it has a few articles ready to write anyway.

A quick look a line profile shows:
- 4.5 second total (CPU time)
- 11% opening the file (115 calls so some duplicates)
- 9% on sabnzbd.ArticleCache.load_article(article)
- 79% writing

So for instance files could be kept open and articles could have a weak reference to the data in the cache avoiding the dictionary lookup.

---

I think the in-order writing is desirable, but @animetosho raises a good point about what to do when the cache is full.

The current overflow logic just gives the bottlenecked disk even more work to do writing to temporary files which will eventually be re-read.

The downloader does try and slow down if there are many things waiting to be written to disk, I havent really looked at it properly to see how well it works but I think potentially if the cache is full downloading should wait until it's all flushed to disk and never write to temporary files.

I also have a PR in progress that's going to combine the download and decode operations which will cut out some operations and maybe help CPU cache hits 🤷

**Reply by puzzledsab** [reply]
I've been thinking a lot about improving the mechanism for finding the next article to fetch but it's difficult to improve because you have to take into account that there can be many servers. The 20 article list is more like a lock, which locks those articles to a particular server by setting the fetcher attribute until it has tried to download them. If it couldn't download them then the other servers will have to try. If there are two or more servers with the same priority then they will grab blocks of 20 articles each. If one fails then the other one has to effectively go back and try the articles it skipped while they were locked.

I changed the `TryList` to be a set once. It helps when several servers have to try the same article but usually it's fetched by the first so it doesn't matter much. Using normal attribute assignment (`dict_["bytes"] = self.bytes`) in `article.__getstate__()` instead of `getattr()` had a bigger impact because it's one of the most frequently called functions and it's about 30% faster. I'm not sure how much it means compared to all the other code. I had stashed it and then I deleted it by mistake and forgot about it.

The cache should no longer overflow as easily as it did in your first artificial tests. I believe the problem then was that they would appear to transfer a single huge file. SABnzbd would never add the same file to the assembler queue more than once so the download speed reduction would never kick in. I changed it so that the number of articles it would wait before it was added to the queue would be 5% of the total cache. Now it will wait 71 articles with a 1GB cache instead of the previously fixed number of 35.

I also changed it so that it always adds it to the queue. If the assembler can't keep up then the queue will normally fill up and the download speed will slow down long before the cache is full. It should now only overflow if there are a lot of missing articles that prevent the assembler from continuing to write the file until they have been checked by all servers.

I have tried writing sparse files in #2459 but I frequently end up with one or two errors according to the par2 check. The same nzbs download correctly with the develop version.

**Reply by mnightingale** [reply]
It's certainly a complicated and confusing, there are about three different get_article or get_articles functions in different places.
I did have a look at it but couldn't think of a more efficient way to structure it

Speaking of `TryList` I noticed every class inheritting from it shares the same global lock, I'm not sure if that is by design or an oversight.
I have noticed the state calls showing up in profiling a few times, are they for saving the queue state to disk? does that happen often?

Since the original post I have noticed the cache overflow behaviour is much better.

I'll take a look at #2459 when I can and see if I can find why it ends up incorrect sometimes, I'm just trying to figure out a race condition with combining download & decode.

**Reply by Safihre** [reply]
Regarding the Try List lock, it's a single lock because there is rarely need for a lock, only the queue reads and writes most of the time. Sometimes the decoder, so it's fine to just have one.

**Comment by puzzledsab** [comment]
We could do what we do several other places to reduce useless calls. Add a `next_assemble_time` to nzf which is a timestamp set for instance 0.2 seconds into the future after each full assemble run. Then we skip the assemble unless current time is higher or the `file_done` flag is true.

**Reply by Safihre** [reply]
@puzzledsab feel free to make a proposal PR :)

PS: I did not forget about your mypyc PR, I want to look into seeing how we can incorporate it.

**Comment by thezoggy** [comment]
looking at logs with latest dev (4.2.0a2), was curious why article cache doesnt seem to use too much and recalled this discussion. in logs i see "Assembler trigger = 72" what is the 72 in regards?

**Reply by Safihre** [reply]
After 72 articles we trigger data to be written to disk (if haven't gotten all articles of a file), it depends on cache size since larger caches can hold more data and write in more continuous larger bursts which is more effecient.
However, 72 articles is around 50MB so probably it just writes every Rar it gets since they'll all be smaller than that.

**Reply by thezoggy** [reply]
I was telling someone on usenet about how usenet parts are usually 100M or less and how we did stuff with article cache that you normally only see limited use because of it.. With my cache set to 1G, seems like it would always trigger at 72. I changed article cache to 4G and now I see that value getting to 244 now. Hard to really see what connections are working on since cache value is on a different tab than connections..   but do we flush to disk per file or really once we hit the article cache/space threshold which could spread across multiple files?

**Reply by Safihre** [reply]
That numerical limit is per file, not total.
If we exceed the total limit we start writing the raw data to disk in seperate files instead of as part of the final file. Very slow.

**Comment by thezoggy** [comment]
Wanted to bring this thread back up as some time has elapsed and was curious about what other optimizations can we do... specifically to deal with increasing internet connection speed + people using more connections... which means the strain on sab incomplete folder rears its head if the cache->disk can't keep up.

goal: how to reduce occurrences of:
"Delayed - # seconds - Assembler queue: ##"

Please correct me if I'm wrong but this is what I currently understand the process once a nzb is queued up.
Sab gets data from usenet service provider for an article, decodes it to store in memory, we assemble and flushed it to disk in the nzbjob/admin folder.  We do NOT wait until all articles for the part is decoded before flushing to disk. Which means we do not write things sequentially as we try to get stuff out of memory faster to reduce that footprint. 

some random thoughts about the article cache -> disk
- Could we yield improvements doing concurrent tasks.. (multiprocessing/threading/asyncio) not blocking?
- Doing batched operations to help alleviate when other tasks are also doing io (ex post-processing / db updates)?
- Is there more parts of this that we can do over in sabctools ?

Has anyone played with python 3.13 GIL disabled mode to see if it makes a difference for this work?
-- py3.13 was supposed to be out a few days ago,  but got pushed back a week due to speed regression.

**Reply by Safihre** [reply]
> Which means we do not write things sequentially as we try to get stuff out of memory faster to reduce that footprint.

We do write in sequence only, we did not end up actually committing the code that would allow non-sequential data writing.

We could easily make the Assembler also multi-threaded by adding a simple queue and some logic to prevent having 2 Assemblers open the same file. I already have a branch somewhere that does that. Because the actual `fout.write()` and the `crc32` calculations release the GIL so multiple writes could occur concurrently. Especially with the free threaded (no-GIL) version of 3.13.
However, in my experience, we see more often people with actually slow disks than the Assembler being at it's limit.
What's the point of trying to write with 2 or more streams at the same time, if the disks can't keep up anyway?

Do you really often have people where the writing to disk is really the problem?

I have not tried 3.13 yet, because pywin32 isn't release for it yet so it would really be a pain for me to get it running on Windows.

**Reply by thezoggy** [reply]
i do see that pywin32 master repo does show that it builds under 3.13
https://github.com/mhammond/pywin32/blob/main/CHANGES.txt
and per their repo that just waiting on a new release: github.com/mhammond/pywin32/issues/2367

i see cryptography -> cffi  while has py3.13 support does not have free-threaded python support:
from v1.17.0 notes
>    Free-threaded CPython builds (i.e. python3.13t and the 3.13t ABI) are not currently supported.

**Reply by thezoggy** [reply]
> > Which means we do not write things sequentially as we try to get stuff out of memory faster to reduce that footprint.
> 
> We do write in sequence only, we did not end up actually committing the code that would allow non-sequential data writing.
> 
> We could easily make the Assembler also multi-threaded by adding a simple queue and some logic to prevent having 2 Assemblers open the same file. I already have a branch somewhere that does that. Because the actual `fout.write()` and the `crc32` calculations release the GIL so multiple writes could occur concurrently. Especially with the free threaded (no-GIL) version of 3.13. However, in my experience, we see more often people with actually slow disks than the Assembler being at it's limit. What's the point of trying to write with 2 or more streams at the same time, if the disks can't keep up anyway?
> 
> Do you really often have people where the writing to disk is really the problem?
> 
> I have not tried 3.13 yet, because pywin32 isn't release for it yet so it would really be a pain for me to get it running on Windows.

what about those that dont just use a single disk but have a pool of disks?
internet speeds are growing and those that have fast connections are prob going to have multiple drives to download to. 

speaking of, prob time to refresh our test nzb and include something larger than 10G

**Reply by puzzledsab** [reply]
I don't really see a good way to use more than 2 paths like it does to day (temp and download). Temp could be split into article cache and rar files but the cached articles aren't supposed to be written temporarily except for when shutting down SAB. If they're written under normal use then it's usually because the temp disk is too slow and the cache fills up because the assembler can't write fast enough. A better way to solve that would be to set a speed limit.

The rar files for one nzb probably have to be in the same dir so the only way I can think of to use more disks would be to download two nzbs at the same time and have two temp dirs that are alternated between. Alternatively the user could set up RAID-0 or similar.

**Reply by thezoggy** [reply]
I agree, I'm guessing just doing a pool/raid setup is prob ideal for those that are just running into performance limitations that the single disk can't keep up..  of course doing things like turning off direct unpack/pause during post-processing to also help manage for ongoing queues besides using a different drive for complete folder (or even different drives for different category if needed) so the io is segmented out when needed.  

My note from 2w ago was more of fallout of upgrading to newer internet and trying to figure out where the bottleneck lies/what was holding me back. Since then, I've learned that having my incomplete drive using XFS instead of BTRFS, and sab docker that used arch instead of alpine as their base os resulted in much more constant performance and higher speeds. I'm guessing there must be some suboptimal things with musl over glibc with how we are using things.

**Reply by mnightingale** [reply]
I expect free-threading will give opportunity for the greatest performance gain, I'm not sure how much if any of SAB will need changes to be thread safe. It may make sense to split downloading/decoding over more threads. Stable builds of it are probably years away and won’t be considered being made the default build for likely years after that.

Focussing on specifically possible optimisations we could make to the assembler; I can’t think of anything particularly suboptimal, usually it opens a file once and always writes sequentially.
Maybe it could struggle more at 500MB/s+ if it needs to write to multiple files but I haven’t analysed it recently.

A few related areas which could be worth considering are:

- We decided against the truncating and writing to correct offsets, I'd suggest at least consider the offsets/seek part as I believe it has some advantages for par2 recovery.
- Currently we download into a reused buffer per connection, because it relieves pressure on the garbage collector.
However, each decode allocates a new bytearray. I'm not sure if we could rework this; I'm not familiar with it but it may be worth investigating something with mmap and just decode straight to the file and avoid the extra allocation/copy - basically eliminates some of the assembler. We'd get the CRC mismatch after writing so need to either leave invalid data or clear it out after.
- Incremental decoding; currently downloads ~800KB and then decodes the whole thing, may be some performance to gain to do this incrementally; I'm thinking CPU cache hits and overall better use of cycles. Buffer per connection could decode in place then copy after or see previous bullet.

Not assembler related:

- I suspect the largest speed limiter is latency any how chatty the protocol is, the spec mentions that clients may use pipelining.
So, send requests for several articles per server at once, and as they are read request more even though you have others still to read.
If it works the result should be that the connection spends less time idle waiting for data and fewer connections maybe required to achieve max speed.
Probably little impact for many connections and lower end of latency but would be interesting USA <=> EU servers.
Also see head-of-line blocking, I’m not sure maybe just need to limit requests to what will fit in one packet or chunk them and request several at a time, read them all, then request several more.

I've tried to look at implementing a few of these, but they would require quite substantial changes.
